source/mir/random/nonuniform.d

```diff
@@ -0,0 +1,246 @@
+
+References:
+    Marsaglia, George, and Wai Wan Tsang. "The ziggurat method for generating random variables."
+    Journal of statistical software 5.8 (2000): 1-7.
+*/
+struct Ziggurat(T, string _fallback, UIntType, size_t numberOfBlocks, bool bothSides)
+    if (isNumeric!T)
+{
+
+private:
+
+    /// monotone decreasing probability density function
+    /// constants don't matter
+    T function(T x) pdf;
+    T function(T x) invPdf;
+
+    /// precalculate difference x_i / x_{i+1}
+    T[numberOfBlocks] xDiv;
+    /// precalculate scaling to R.max for x_i
+    T[numberOfBlocks] xScaled;
+    /// precalculate pdf value for x_i
+    T[numberOfBlocks] fs;
+
+    // mask to use to get the first log_2 k bits (=k - 1)
+    enum size_t kMask = numberOfBlocks - 1;
+
+    /// left-point of the right-most block
+    T rightEnd;
+
+    /// area of every column
+    T blockArea;
+
+public:
+    /**
+    Params:
+        pdf = probability density function
+        invPdf = inverse probability density function
+        rightEnd = left point of the right-most block
+        blockArea =  area of every column
+    */
+    this(T function(T x) pdf, T function(T x) invPdf, T rightEnd, T blockArea)
+    {
+        this.pdf = pdf;
+        this.invPdf = invPdf;
+        this.rightEnd = rightEnd;
+        this.blockArea = blockArea;
+
+        // scale factor to the range of UIntType
+        T maxT = bothSides ? UIntType.max / 2 : UIntType.max;
+
+        auto xn = rightEnd; // Next x_{i+1}
+        auto xc = xn; // Current x_i
+        auto fn = pdf(xn);
+
+        // k_i = floor(2^32 (x_{i-1} / x_i))
+        xDiv[0] = (xn * fn / blockArea) * maxT;
+        xDiv[1] = 0;
+
+        // w_i = 0.5^32 * x_i
+        xScaled[0]     = (blockArea / fn) / maxT;
+        xScaled[$ - 1] = xn / maxT;
+
+        // f_i = f(x_i)
+        fs[0]     = T(1);
+        fs[$ - 1] = fn;
+
+        for (auto i = xScaled.length - 2; i >= 1; i--)
+        {
+            xn = invPdf(blockArea / xn + fn);
+            xDiv[i + 1] = (xn / xc) * maxT;
+            fn = fs[i] = pdf(xn);
+            xScaled[i] = xn / maxT;
+            xc = xn;
+        }
+    }
+
+    /// samples a value from the discrete distribution
+    T opCall() const
+    {
+        import std.random : rndGen;
+        return opCall(rndGen);
+    }
+
+    /// samples a value from the discrete distribution using a custom random generator
+    T opCall(RNG)(ref RNG gen) const
+        if (typeof(gen.front).sizeof == UIntType.sizeof)
+    {
+        import std.random : uniform;
+        import std.traits : Signed;
+        import std.range : ElementType;
+        import std.math : abs;
+
+        // TODO: option for inlining
+        for (;;)
+        {
+            static if (bothSides)
+                Signed!(ElementType!RNG) u = gen.front;
+            else
+                auto u = gen.front;
+
+            gen.popFront();
+
+            // TODO: this is a bit biased
+            size_t i = u & kMask;
```

Comment:
Yes resolving whether we can reuse the random bits or need to generate another 1/4 random variable (one of the optimizations described in later papers), shouldn't matter so much :/